lib/LanguageServerBridge/LanguageServerBridgeExtension.php

```diff
@@ -25,11 +29,28 @@ public function configure(Resolver $schema)
     public function load(ContainerBuilder $container)
     {
         $container->register(LocationConverter::class, function (Container $container) {
-            return new LocationConverter($container->get(LanguageServerExtension::SERVICE_SESSION_WORKSPACE));
+            return new LocationConverter(
+                $container->get(TextDocumentLocator::class)
+            );
         });
 
         $container->register(TextEditConverter::class, function (Container $container) {
             return new TextEditConverter();
         });
+
+        $container->register(FilesystemWorkspaceLocator::class, function (Container $container) {
+            return new FilesystemWorkspaceLocator();
+        });
+
+        $container->register(WorkspaceTextDocumentLocator::class, function (Container $container) {
+            return new WorkspaceTextDocumentLocator($container->get(LanguageServerExtension::SERVICE_SESSION_WORKSPACE));
+        });
+
+        $container->register(TextDocumentLocator::class, function (Container $container) {
+            return new ChainDocumentLocator([
+                $container->get(WorkspaceTextDocumentLocator::class),
+                $container->get(FilesystemWorkspaceLocator::class)
+            ]);
```

Comment:
@BladeMF this is the service which can be used to either get from the workspace all the FS.

Reply:
or with less typos: to retrieve a document from either the workspace or the filessytem

Reply:
Which projects changed with this? TextDocument?

Reply:
yes just TextDocument

Reply:
I need help setting this up for testing. How do I setup an instance of that `WorkspaceTextDocumentLocator` to pass to my renamers?

Reply:
you can use the `InMemoryTextDocumentLocator(TextDocument[])`

Reply:
otherwise something like https://github.com/phpactor/language-server-phpactor-extensions/blob/8e6f7d85dbcff3d812e54508228db6eb11ab8d32/tests/LanguageServerReferenceFinder/Unit/Handler/GotoDefinitionHandlerTest.php#L28-L35

Reply:
Oh, nice, thanks!